presto-native-execution/presto_cpp/main/PrestoServer.cpp

```diff
@@ -1426,9 +1426,56 @@ void PrestoServer::populateMemAndCPUInfo() {
   });
   RECORD_METRIC_VALUE(kCounterNumQueryContexts, numContexts);
   cpuMon_.update();
+  checkOverload();
   **memoryInfo_.wlock() = std::move(memoryInfo);
 }
 
+void PrestoServer::checkOverload() {
+  auto systemConfig = SystemConfig::instance();
+
+  const auto overloadedThresholdMemBytes =
+      systemConfig->workerOverloadedThresholdMemGb() * 1024 * 1024 * 1024;
+  if (overloadedThresholdMemBytes > 0) {
+    const auto currentUsedMemoryBytes = (memoryChecker_ != nullptr)
+        ? memoryChecker_->cachedSystemUsedMemoryBytes()
+        : 0;
+    const bool isMemOverloaded =
+        (currentUsedMemoryBytes > overloadedThresholdMemBytes);
+    if (isMemOverloaded) {
+      LOG(WARNING) << "Server memory is overloaded. Currently used: "
+                   << velox::succinctBytes(currentUsedMemoryBytes)
+                   << ", threshold: "
+                   << velox::succinctBytes(overloadedThresholdMemBytes);
+    } else if (isMemOverloaded_) {
+      LOG(INFO) << "Server memory is no longer overloaded. Currently used: "
+                << velox::succinctBytes(currentUsedMemoryBytes)
+                << ", threshold: "
+                << velox::succinctBytes(overloadedThresholdMemBytes);
+    }
+    RECORD_METRIC_VALUE(kCounterOverloadedMem, isMemOverloaded ? 100 : 0);
```

Comment:
Curious why is 0 and 100, but not a range of values in between. or why not just 0 and 1.

Reply:
@amitkdutta 
Because when the counters backend starts averaging the signal over some time period, say 60 seconds the 1 will turn into 0 and we won't see any signal.
100 however won't have such issue and will be easy to reason about how much % of the period the system was overloaded if the number is below 100.

Reply:
We are adding more and more stuff to PrestoServer now. It will be easier to read and maintain if we move these logic to a separate class maybe called SystemMonitor. We can start from this one and maybe move a couple of stuff from current PrestoServer there, as well.